src/ferc_xbrl_extractor/datapackage.py

```diff
@@ -346,44 +351,34 @@ def construct_dataframe(self, instance: Instance) -> pd.DataFrame:
         Args:
             instance: Parsed XBRL instance used to construct dataframe.
         """
-        # Filter contexts to only those relevant to table
-        contexts = instance.get_facts(self.instant, self.axes)
-
-        # Get the maximum number of rows that could be in table and allocate space
-        max_len = len(contexts)
-
-        # Allocate space for columns
-        df = {
-            field.name: pd.Series([pd.NA] * max_len, dtype=FIELD_TO_PANDAS[field.type_])
-            for field in self.schema.fields
-        }
-
         # Loop through contexts and get facts in each context
         # Each context corresponds to one unique row
-        for i, (context, facts) in enumerate(contexts.items()):
-            if self.instant != context.period.instant:
-                continue
-
-            # Construct dictionary to represent row which corresponds to current context
-            row = {fact.name: fact.value for fact in facts if fact.name in df}
-
-            # If row is empty skip
-            if row:
-                # Use context to create primary key for row and add to dictionary
-                row.update(context.as_primary_key(instance.filing_name))
-
-                # Loop through each field in row and add to appropriate column
-                for key, val in row.items():
-                    # Try to parse value from XBRL if not possible it will remain NA
-                    try:
-                        df[key][i] = self.columns[key](val)
-                    except ValueError:
-                        self.logger.warning(
-                            f"Could not parse {val} as {self.columns[key]} in column {key}"
-                        )
-
-        # Create dataframe and drop empty rows
-        return pd.DataFrame(df).dropna(how="all")
+        df = {}
+        for column in self.data_columns:
+            df[column] = {
+                c_id: self.columns[column](fact.value)
+                for c_id, fact in instance.get_facts(
+                    self.instant, column, self.schema.primary_key
+                ).items()
+            }
+
+        # Create dataframe indexed by context ID
+        df = pd.DataFrame(df)
+
+        # Expand context ID to contain columns for date, entity ID, and any axes
+        primary_key = pd.DataFrame(
+            {
+                c_id: instance.contexts[c_id].as_primary_key(
+                    instance.filing_name, self.axes
+                )
+                for c_id in df.index
+            }
+        )
+
+        # Join on context ID
+        df = primary_key.T.join(df)
```

Comment:
If you want to make this dataframe without having to `T`ranspose it, you can do `pd.DataFrame.from_dict({...}, orient="index")`.